### ISSUE TITLE
TCP relay block when first packet includes payload

### ISSUE THREAD

**Issue by cute** [shadowsocks/go-shadowsocks2#37]
when I copy left buffer manually at first, it is woking.

```
func relay(left, right net.Conn) (int64, int64, error) {
	type res struct {
		N   int64
		Err error
	}

	ch := make(chan res)

	buf := make([]byte, 1024)
	for {
		nread, _ := left.Read(buf)
		if nread > 0 {
			right.Write(buf[0:nread])
		}
		if nread != 1024 {
			break
		}
	}

	go func() {
		n, err := io.Copy(right, left)
		right.SetDeadline(time.Now()) // wake up the other goroutine blocking on right
		left.SetDeadline(time.Now())  // wake up the other goroutine blocking on left
		ch <- res{n, err}
	}()

	n, err := io.Copy(left, right)
	right.SetDeadline(time.Now()) // wake up the other goroutine blocking on right
	left.SetDeadline(time.Now())  // wake up the other goroutine blocking on left
	rs := <-ch

	if err == nil {
		err = rs.Err
	}
	return n, rs.N, err
}
```

**Comment by riobard** [comment]
Sorry I didn't understand the problem. What happened here? And what's the step to reproduce it?

**Comment by cute** [comment]
@riobard 

go-shadowsocks2 client use socks5 proxy protocol, send address to server at first,
then send request header to server.
but my client use http proxy protocol, send address and request header to server.

**Comment by riobard** [comment]
Are you using go-shadowsocks2 as an HTTP proxy?

**Comment by cute** [comment]
no, i write a shadowsocks client that provides http proxy.

**Comment by riobard** [comment]
Sorry I'm confused. Is this issue a bug report? What was wrong here?

**Comment by arthurkiller** [comment]
@cute  Maybe you'd better read the comments about the io.Copy() 

>// Copy copies from src to dst until either EOF is reached
// on src or an error occurs. It returns the number of bytes
// copied and the first error encountered while copying, if any.
//
// A successful Copy returns err == nil, not err == EOF.
// Because Copy is defined to read from src until EOF, it does
// not treat an EOF from Read as an error to be reported.
//
// If src implements the WriterTo interface,
// the copy is implemented by calling src.WriteTo(dst).
// Otherwise, if dst implements the ReaderFrom interface,
// the copy is implemented by calling dst.ReadFrom(src).

**Comment by cute** [comment]
@riobard 
shadowsocks tcp protocol is ` [target address][payload]`, http request through local proxy:

local socks5 proxy send data to server: 

>write [target address]
>write [payload]
>read

local http proxy send data to server: 
>write [target address]_[payload]_
>read

If address and payload is sent together, server can not process the data correctly.

**Comment by riobard** [comment]
I think I understand the problem now. Do you have a code snippet to reproduce the bug?

**Comment by cute** [comment]
@riobard what's your email? I want to send you an iOS app promotion code.

**Comment by riobard** [comment]
me at riobard dot com

**Comment by riobard** [comment]
Got the code. What's the step to produce the bug?

**Comment by riobard** [comment]
Problem identified: this is actually caused by a bug in the implementation of AEAD stream reader. The fix has been committed into my fork here https://github.com/riobard/go-shadowsocks2/commit/de996c889eae0ad0356d654eff7b2ff7aa489096 and a new test release is created at https://github.com/riobard/go-shadowsocks2/releases/tag/v0.0.9

**Comment by bigtan** [comment]
@riobard seem still not work with shadowrocket

**Comment by bigtan** [comment]
fixed through https://github.com/riobard/go-shadowsocks2/pull/8

**Comment by riobard** [comment]
Fixed by #40
